### DISCUSSION TITLE
Use compile-time cols in `point_simplex_squared_distance`

### DISCUSSION THREAD

**Discussion by svenpilz** [libigl/libigl#1959] [Ideas]
```cpp
template <
  int DIM,
  typename Derivedp,
  typename DerivedV,
  typename DerivedEle,
  typename Derivedsqr_d,
  typename Derivedc>
IGL_INLINE void igl::point_simplex_squared_distance(
  const Eigen::MatrixBase<Derivedp> & p,
  const Eigen::MatrixBase<DerivedV> & V,
  const Eigen::MatrixBase<DerivedEle> & Ele,
  const typename DerivedEle::Index primitive,
  Derivedsqr_d & sqr_d,
  Eigen::MatrixBase<Derivedc> & c)
{
  // Use Dynamic because we don't know Ele.cols() at compile time.
  Eigen::Matrix<typename Derivedc::Scalar,1,Eigen::Dynamic> b;
  point_simplex_squared_distance<DIM>( p, V, Ele, primitive, sqr_d, c, b );
}
```
(From: [point_simplex_squared_distance](https://github.com/libigl/libigl/blob/79dc4f683877a8bb4eaf63620437a5454f6c47c1/include/igl/point_simplex_squared_distance.cpp#L153))

Currently it uses `Eigen::Matrix<…, Eigen::Dynamic> b;`, would it make sense to allow for `Eigen::Matrix<…, DerivedEle::ColsAtCompileTime> b;`?

From Eigen's documentation
> The number of columns at compile-time. This is just a copy of the value provided by the Derived type. If a value is not known at compile-time, it is set to the Dynamic constant. 

As I read it, it would still fall back to `Eigen::Dynamic`, but could use `ColsAtCompileTime` which then would safe a memory allocation in https://github.com/libigl/libigl/blob/79dc4f683877a8bb4eaf63620437a5454f6c47c1/include/igl/point_simplex_squared_distance.cpp#L132, which could be expensive when used for a large number of query points from `AABB`.

I found this while profiling an Emscripten build with enabled multi-threading: https://github.com/emscripten-core/emscripten/discussions/15914.

**Comment by alecjacobson** [comment]
I think this is a good idea. I wonder if this should be excised and re-written though. The templating on dimension for AABB has been a real pain and particularly when hitting this code point.

**Comment by svenpilz** [comment]
We made some further testing and may found an additional bottleneck.

So far we have: https://github.com/libigl/libigl/compare/main...svenpilz:reduceDynamicAllocations?expand=1

Would this be interesting as a pull-request for you?

Edit: Added as https://github.com/libigl/libigl/pull/2001.
